### PR TITLE
Challenge 1- Number Guessing Game From the Updated Branch

### PR DESCRIPTION
**Pull request by malakbattat**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [ ] All tests pass
- [x] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [ ] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Rouaa93 — COMMENTED** [review]
Good job, Malak! The code is organized and the tests are clear. I just have a few suggestions:
	1.	Try adding tests for edge cases, like the smallest and largest guesses.
	2.	Add more details in the docstrings about what inputs the functions expect and what errors they might raise.
	3.	Make sure the file name matches the main function name for consistency.

Great work @malakbattat overall !!🤩👌

File: solutions/tests/test_ch1_malak.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            self.assertEqual(
+                message,
+                "Invalid input! Defaulting to Medium (1 to 50) with 10 attempts.",
+            )
+
+
+class TestNumberGuessingGame(unittest.TestCase):
+    """Unit tests for the `number_guessing_game` function."""
+
+    def test_number_guessing_game(self):
```
Comment (abdoalsir) [inline solutions/tests/test_ch1_malak.py:59]: It is generally good to limit each unit test to a single assertion to ensure clarity and better isolation of test failures. Consider splitting the test into multiple smaller tests, each verifying one specific behavior.
Also, when running the test file it prompts the tester to play the game rather than testing the functions, so make sure to review that as well

File: solutions/solution_ch1_malak.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+"""
+This module implements a number guessing game where players select a difficulty
+level and attempt to guess a randomly generated number within a limited number of attempts.
+The difficulty levels determine the number of attempts allowed.
+"""
+
+import random
+
+
+def choose_difficulty() -> tuple[int, int, str]:
```
Comment (abdoalsir) [inline solutions/solution_ch1_malak.py:10]: It looks like this function is missing doctests. Adding doctests can be a great way to demonstrate how the function is expected to be used and help verify its correctness. This also makes it easier for others to understand and maintain the code. Consider adding some examples to illustrate its usage.

File: solutions/tests/test_ch1_malak.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+"""
+Unit tests for the choose_difficulty function from the number guessing game module.
```
Comment (abdoalsir) [inline solutions/tests/test_ch1_malak.py:2]: The function docstring is well-written, but it seems the module is missing a docstring. It can provide an overview of the module's purpose, its main functionalities, and any relevant context. This helps other developers understand the module at a glance. Consider adding one to enhance clarity and completeness.

**Review by abdoalsir — CHANGES_REQUESTED** [review]
Your game idea is creative and the code looks solid and with some minor improvements, it would be perfect.
**Final thoughts on the code:**
- Consider separating the input handling logic and game mechanics more clearly by splitting functionality into distinct functions or classes, as you've suggested.
- Finally, You can go over the checklist again and review the code yourself as well, and I believe you will also have some useful points of view on the code for better implementation.

**Comment by malakbattat** [comment]
Thank you @Rouaa93 and @abdoalsir for the insightful reviews, I've put them into consideration while editing for my code.

**Comment by malakbattat** [comment]
I've actually refactored the code to remove user input and output, focusing on function arguments and return values instead. Updated docstrings to describe behavior, added doctests, and ensured clean, modular, and testable code. Test file uses absolute imports and has clear, concise unit tests.
